src/enclave/rpcendpoint.h

```diff
@@ -34,39 +35,98 @@ namespace enclave
       session_id(session_id)
     {}
 
+    auto split_actor_and_method(const std::string& actor_method)
+    {
+      const auto split_point = actor_method.find_last_of('/');
+      const auto actor = actor_method.substr(0, split_point);
+      const auto method =
+        actor_method.substr(split_point + 1, actor_method.size());
+
+      // NB: If the string does not contain '/', then both actor and method will
+      // contain the entire string
+      return std::make_pair(actor, method);
+    }
+
     bool handle_data(const std::vector<uint8_t>& data)
     {
-      if (!handler)
+      LOG_DEBUG_FMT(
+        "Entered handle_data, session {} with {} bytes",
+        session_id,
+        data.size());
+
+      const SessionContext session(session_id, peer_cert());
+      RPCContext rpc_ctx(session);
+
+      rpc_ctx.raw = data;
+
+      auto [success, rpc] = jsonrpc::unpack_rpc(data, rpc_ctx.pack);
+
+      if (!success)
+      {
+        send(jsonrpc::pack(rpc, rpc_ctx.pack.value()));
+        return true;
+      }
+      LOG_TRACE_FMT("Deserialised");
+
+      parse_rpc_context(rpc_ctx, rpc);
```

Comment:
Seeing as you are doing the same here, what about unpacking first, and if everything is OK, passing `SessionContext`, `data` and `rpc` into `parse_rpc_context` and have that create and return your `RPCContext` object. That way you don't have to repeat this code.

Reply:
I'd prefer to keep the current format, as in general the context parsing is multi-stage - populating the context piece-by-piece as we peel off additional layers of wrapping. The JSON-RPC case doesn't show this very well since we only go through 3 stages (empty context tied to a session, detecting the packing format during `unpack_rpc`, extracting all other fields in `parse_rpc_context`), but its useful to be able to return appropriate errors between each stage. I want to keep this pattern, and if anything pull more stages out (extracting the signature perhaps), rather than storing features temporarily to build the finished context.

This causes some code repetition, but I think that's inevitable to support our tests. `handle_data` is the real parsing function, returning detailed error messages to a client if the parse fails. `make_rpc_context` does something similar for the tests, so they can call `process` directly.